src/Libraries/Microsoft.Extensions.AI/ToolReduction/EmbeddingToolReductionStrategy.cs

```diff
@@ -0,0 +1,330 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Numerics.Tensors;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI;
+
+#pragma warning disable IDE0032 // Use auto property, suppressed until repo updates to C# 14
+
+/// <summary>
+/// A tool reduction strategy that ranks tools by embedding similarity to the current conversation context.
+/// </summary>
+/// <remarks>
+/// The strategy embeds each tool (name + description by default) once (cached) and embeds the current
+/// conversation content each request. It then selects the top <c>toolLimit</c> tools by similarity.
```

Comment:
I wonder if there is a hybrid that can reduce tools, but then also expose a "get more tools" function where the model could describe the sort of tool it wanted and we could use that to give it back more tools that have embedding similarity to its request?  Don't want to feature creep this, just trying to think of alternatives.

I like that this one is hands free but but I can imagine that it can hit situations where it never gives the model the chance to call what might be the best tool.

I like that the grouping ensures that all tools are made available, but it requires explicit grouping by the caller.

Reply:
Maybe what we could do is change `MessagesEmbeddingTextSelector` to be a `Func<IEnumerable<ChatMessage>, Task<string>>`. Then if you wanted to have the model separately describe the tool it needs, you could asynchronously have a chat client analyze the conversation and generate a short string describing the type of tool that would be helpful. Then that string is what gets used when searching for "similar" tools.

Reply:
Just implemented the above ^